File: go-server/internal/handler/services.go
package handler

import (
	"bytes"
	"encoding/json"
	"net/http"
	"strconv"
	"sync"
	"time"

	"github.com/nats-io/nats.go"

	"nats-explorer/internal/connection"
)

type ServicesHandler struct {
	Store *connection.Store
}

const (
	defaultDiscoverWait = 1200 * time.Millisecond
	maxDiscoverWait     = 10 * time.Second
	maxDiscoverResults  = 500
)

// discoverVia publishes a $SRV request and collects every reply that arrives
// within the wait window. Replies are appended under a mutex because the
// subscription callback runs on the NATS client goroutine.
func (h *ServicesHandler) discoverVia(w http.ResponseWriter, r *http.Request, subject string) {
	nc, err := h.Store.GetNC(connIDFromRequest(r))
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	wait := defaultDiscoverWait
	if ms, err := strconv.Atoi(r.URL.Query().Get("waitMs")); err == nil && ms > 0 {
		wait = time.Duration(ms) * time.Millisecond
		if wait > maxDiscoverWait {
			wait = maxDiscoverWait
		}
	}

	var mu sync.Mutex
	results := make([]json.RawMessage, 0)

	inbox := nc.NewRespInbox()
	sub, err := nc.Subscribe(inbox, func(msg *nats.Msg) {
		// Without a responder the server answers with an empty 503 message.
		// It carries no service, and passing it on would put a null into the
		// list the browser reads.
		if !isServiceReply(msg) {
			return
		}
		mu.Lock()
		defer mu.Unlock()
		if len(results) < maxDiscoverResults {
			results = append(results, json.RawMessage(append([]byte(nil), msg.Data...)))
		}
	})
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	defer sub.Unsubscribe()

	if err := nc.PublishRequest(subject, inbox, nil); err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	nc.Flush()

	select {
	case <-time.After(wait):
	case <-r.Context().Done():
		return
	}

	mu.Lock()
	out := make([]json.RawMessage, len(results))
	copy(out, results)
	mu.Unlock()

	writeJSON(w, out)
}

// isServiceReply reports whether a reply carries a service description: a
// JSON object, not a status message and not empty.
func isServiceReply(msg *nats.Msg) bool {
	if msg.Header.Get("Status") != "" {
		return false
	}
	data := bytes.TrimSpace(msg.Data)
	return len(data) > 0 && data[0] == '{' && json.Valid(data)
}

func (h *ServicesHandler) Discover(w http.ResponseWriter, r *http.Request) {
	h.discoverVia(w, r, "$SRV.INFO")
}

func (h *ServicesHandler) Stats(w http.ResponseWriter, r *http.Request) {
	h.discoverVia(w, r, "$SRV.STATS")
}

func (h *ServicesHandler) Ping(w http.ResponseWriter, r *http.Request) {
	h.discoverVia(w, r, "$SRV.PING")
}

File: server/src/routes/services.ts
import { Router } from 'express';
import { StringCodec } from 'nats';
import { connectionStore } from '../nats/connection-manager.js';

export const servicesRouter: Router = Router();
const sc = StringCodec();

// Discover all services via $SRV.INFO
servicesRouter.get('/services', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const nc = connectionStore.getConnection(connId);
    if (!nc) return res.status(400).json({ error: 'Not connected' });

    const services: any[] = [];
    const sub = nc.subscribe('$SRV.INFO', { max: 50, timeout: 3000 });

    // Request service info
    nc.publish('$SRV.INFO', undefined);

    try {
      for await (const msg of sub) {
        try {
          const info = JSON.parse(sc.decode(msg.data));
          services.push(info);
        } catch {}
      }
    } catch {
      // Timeout is expected
    }

    res.json(services);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// Get stats for all services via $SRV.STATS
servicesRouter.get('/services/stats', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const nc = connectionStore.getConnection(connId);
    if (!nc) return res.status(400).json({ error: 'Not connected' });

    const stats: any[] = [];
    const sub = nc.subscribe('$SRV.STATS', { max: 50, timeout: 3000 });

    nc.publish('$SRV.STATS', undefined);

    try {
      for await (const msg of sub) {
        try {
          const stat = JSON.parse(sc.decode(msg.data));
          stats.push(stat);
        } catch {}
      }
    } catch {}

    res.json(stats);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});

// Ping all services via $SRV.PING
servicesRouter.get('/services/ping', async (req, res) => {
  try {
    const connId = req.query.connId as string;
    if (!connId) return res.status(400).json({ error: 'connId required' });
    const nc = connectionStore.getConnection(connId);
    if (!nc) return res.status(400).json({ error: 'Not connected' });

    const pongs: any[] = [];
    const sub = nc.subscribe('$SRV.PING', { max: 50, timeout: 2000 });

    nc.publish('$SRV.PING', undefined);

    try {
      for await (const msg of sub) {
        try {
          const pong = JSON.parse(sc.decode(msg.data));
          pongs.push(pong);
        } catch {}
      }
    } catch {}

    res.json(pongs);
  } catch (err: any) {
    res.status(500).json({ error: err.message });
  }
});
